File: uigesturerecognizer/src/main/java/it/sephiroth/android/library/uigestures/UIGestureRecognizer.java
package it.sephiroth.android.library.uigestures;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Handler;
import android.os.Message;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.Log;
import android.view.MotionEvent;
import android.view.ViewConfiguration;

import java.util.ArrayList;
import java.util.List;

import it.sephiroth.android.library.simplelogger.LoggerFactory;

/**
 * Created by <b>Alessandro Crugnola</b>
 * <p>AndroidGestureRecognizer is an Android implementation
 * of the Apple's UIGestureRecognizer framework. There's not guarantee, however, that
 * this library works 100% in the same way as the Apple version.</p>
 *
 * @author sephiroth
 * @version 1.0.0
 * @see <a href='https://developer.apple.com/reference/uikit/uigesturerecognizer'>
 * https://developer.apple.com/reference/uikit/uigesturerecognizer</a>
 */

@SuppressWarnings ("unused")
public abstract class UIGestureRecognizer implements OnGestureRecognizerStateChangeListener {

    public enum State {
        Possible,
        Began,
        Changed,
        Failed,
        Cancelled,
        Ended
    }

    public interface OnActionListener {
        void onGestureRecognized(@NonNull final UIGestureRecognizer recognizer);
    }

    public static final String VERSION = BuildConfig.VERSION_NAME;
    private static int sId = 0;
    private static boolean sDebug = false;

    static final long LONG_PRESS_TIMEOUT = ViewConfiguration.getLongPressTimeout();
    static final long TAP_TIMEOUT = ViewConfiguration.getTapTimeout();
    static final long DOUBLE_TAP_TIMEOUT = ViewConfiguration.getDoubleTapTimeout();
    static final int TOUCH_SLOP = 8;
    static final int DOUBLE_TAP_SLOP = 100;
    static final int DOUBLE_TAP_TOUCH_SLOP = TOUCH_SLOP;

    private final List<OnGestureRecognizerStateChangeListener> mStateListeners = new ArrayList<>();
    private OnActionListener mListener;
    private State mState;
    private boolean mEnabled;
    private boolean mCancelsTouchesInView;
    private UIGestureRecognizerDelegate mDelegate;
    private Object mTag;
    private long mId;
    private UIGestureRecognizer mOtherRecognizer;
    private final LoggerFactory.Logger logger = LoggerFactory.getLogger(getClass().getSimpleName());

    protected final GestureHandler mHandler;

    public UIGestureRecognizer(@Nullable Context context) {
        mHandler = new GestureHandler();
        mCancelsTouchesInView = true;
        mEnabled = true;
        mId = generateId();
    }

    private long generateId() {
        return sId++;
    }

    @SuppressLint ("HandlerLeak")
    protected final class GestureHandler extends Handler {
        @Override
        public void handleMessage(final Message msg) {
            UIGestureRecognizer.this.handleMessage(msg);
        }
    }

    protected abstract void removeMessages();

    protected final void removeMessages(int... messages) {
        for (int message : messages) {
            mHandler.removeMessages(message);
        }
    }

    protected final boolean hasMessages(int... messages) {
        for (int message : messages) {
            if (mHandler.hasMessages(message)) {
                return true;
            }
        }
        return false;
    }

    protected final void setDelegate(final UIGestureRecognizerDelegate delegate) {
        mDelegate = delegate;
    }

    protected final UIGestureRecognizerDelegate getDelegate() {
        return mDelegate;
    }

    protected final void fireActionEvent() {
        if (null != mListener) {
            mListener.onGestureRecognized(this);
        }
    }

    protected void addOnStateChangeListenerListener(final OnGestureRecognizerStateChangeListener listener) {
        if (!mStateListeners.contains(listener)) {
            mStateListeners.add(listener);
        }
    }

    protected boolean removeOnStateChangeListenerListener(final OnGestureRecognizerStateChangeListener listener) {
        return mStateListeners.remove(listener);
    }

    protected abstract boolean onTouchEvent(MotionEvent event);

    protected abstract void handleMessage(final Message msg);

    /**
     * @param mTag custom object the instance should keep
     * @since 1.0.0
     */
    public void setTag(final Object mTag) {
        this.mTag = mTag;

        if (sDebug) {
            logger.setTag(String.valueOf(mTag));
        }
    }

    /**
     * @return current tag assigned to this instance
     * @since 1.0.0
     */
    public Object getTag() {
        return mTag;
    }

    /**
     * @param mId change the instance id
     */
    public void setId(final long mId) {
        this.mId = mId;
    }

    /**
     * @return the instance id
     * @since 1.0.0
     */
    public static int getId() {
        return sId;
    }

    public void setActionListener(final OnActionListener listener) {
        this.mListener = listener;
    }

    /**
     * @return Returns the number of touches involved in the gesture represented by the receiver.
     * @since 1.0.0
     */
    public abstract int getNumberOfTouches();

    /**
     * @return Returns the X computed as the location in a given view of the gesture represented by the receiver.
     * @since 1.0.0
     */
    public abstract float getCurrentLocationX();

    /**
     * @return Returns the Y computed as the location in a given view of the gesture represented by the receiver.
     * @since 1.0.0
     */
    public abstract float getCurrentLocationY();

    /**
     * @return The current recognizer internal state
     * @since 1.0.0
     */
    public final State getState() {
        return mState;
    }

    protected final void setState(State state) {
        logMessage(Log.INFO, "setState: %s", state);

        final boolean changed = mState != state || state == State.Changed;
        mState = state;

        if (changed) {
            for (OnGestureRecognizerStateChangeListener listener : mStateListeners) {
                listener.onStateChanged(this);
            }
        }
    }

    protected boolean inState(State... states) {
        if (null == states || states.length < 1) {
            return false;
        }
        for (State state : states) {
            if (mState == state) {
                return true;
            }
        }
        return false;
    }

    /**
     * @return True if the recognizer is enabled
     * @since 1.0.0
     */
    public final boolean isEnabled() {
        return mEnabled;
    }

    /**
     * Toggle the recognizer enabled state.
     *
     * @param enabled Set to false to prevent any motion event
     *                to be intercepted by this recognizer
     * @since 1.0.0
     */
    public final void setEnabled(boolean enabled) {
        mEnabled = enabled;
    }

    /**
     * @param other Creates a dependency relationship between the receiver and another gesture recognizer when the objects
     *              are created
     * @see <a href='https://developer.apple.com/reference/uikit/uigesturerecognizer/1624203-require'>
     * https://developer.apple.com/reference/uikit/uigesturerecognizer/1624203-require</a>
     * @since 1.0.0
     */
    public final void requireFailureOf(@Nullable final UIGestureRecognizer other) {
        if (null != mOtherRecognizer) {
            mOtherRecognizer.removeOnStateChangeListenerListener(this);
        }
        this.mOtherRecognizer = other;
    }

    protected final UIGestureRecognizer getRequireFailureOf() {
        return mOtherRecognizer;
    }

    protected final void stopListenForOtherStateChanges() {
        if (null != getRequireFailureOf()) {
            getRequireFailureOf().removeOnStateChangeListenerListener(this);
        }
    }

    protected final void listenForOtherStateChanges() {
        if (null != getRequireFailureOf()) {
            getRequireFailureOf().addOnStateChangeListenerListener(this);
        }
    }

    /**
     * @param value A Boolean value affecting whether touches are delivered to a view when a gesture is recognized
     * @see <a href='https://developer.apple.com/reference/uikit/uigesturerecognizer/1624218-cancelstouchesinview'>
     * https://developer.apple.com/reference/uikit/uigesturerecognizer/1624218-cancelstouchesinview</a>
     * @since 1.0.0
     */
    public final void setCancelsTouchesInView(boolean value) {
        mCancelsTouchesInView = value;
    }

    /**
     * @see UIGestureRecognizer#setCancelsTouchesInView(boolean)
     * @since 1.0.0
     */
    public boolean getCancelsTouchesInView() {
        return mCancelsTouchesInView;
    }

    @Override
    public String toString() {
        return getClass().getSimpleName() + "[state: " + getState() + ", tag:" + getTag() + "]";
    }

    public static void setLogEnabled(boolean enabled) {
        sDebug = enabled;
    }

    void logMessage(int level, String fmt, Object... args) {
        if (!sDebug) {
            return;
        }

        switch (level) {
            case Log.INFO:
                logger.info(fmt, args);
                break;
            case Log.DEBUG:
                logger.debug(fmt, args);
                break;
            case Log.ASSERT:
            case Log.ERROR:
                logger.error(fmt, args);
                break;
            case Log.WARN:
                logger.warn(fmt, args);
                break;
            case Log.VERBOSE:
                logger.verbose(fmt, args);
                break;
            default:
                break;
        }
    }
}

File: uigesturerecognizer/src/main/java/it/sephiroth/android/library/uigestures/UIGestureRecognizer.kt
package it.sephiroth.android.library.uigestures

import android.annotation.SuppressLint
import android.annotation.TargetApi
import android.content.Context
import android.graphics.PointF
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.os.Message
import android.util.Log
import android.view.MotionEvent
import android.view.ViewConfiguration
import java.lang.ref.WeakReference
import java.util.*

/**
 * AndroidGestureRecognizer is an Android implementation
 * of the Apple's UIGestureRecognizer framework. There's not guarantee, however, that
 * this library works 100% in the same way as the Apple version.<br></br>
 * This is the base class for all the UI gesture implementations.
 *
 * @author alessandro crugnola
 * @version 1.0.0
 * @see <a href='https://developer.apple.com/reference/uikit/uigesturerecognizer'>uigesturerecognizer</a>
 */

@Suppress("MemberVisibilityCanBePrivate")
abstract class UIGestureRecognizer(context: Context) : OnGestureRecognizerStateChangeListener {

    private val mStateListeners = Collections.synchronizedList(ArrayList<OnGestureRecognizerStateChangeListener>())

    private var mBeganFiringEvents: Boolean = false

    private val mContextRef: WeakReference<Context> = WeakReference(context)

    private var mNumberOfTouches: Int = 0

    // current ACTION_DOWN location
    protected val mDownLocation = PointF()

    // previous ACTION_DOWN location
    protected val mPreviousDownLocation = PointF()

    // current ACTION_DOWN time
    protected var mDownTime: Long = 0L

    // previous ACTION_DOWN time
    protected var mPreviousDownTime: Long = 0L

    // current event x,y location
    protected val mCurrentLocation = PointF()

    protected val mHandler: GestureHandler = GestureHandler(Looper.getMainLooper())

    protected val isListeningForOtherStateChanges: Boolean get() = null != requireFailureOf && requireFailureOf!!.hasOnStateChangeListenerListener(this)

    internal var delegate: UIGestureRecognizerDelegate? = null

    /**
     * UIGestureRecognizer callback
     */
    var actionListener: ((UIGestureRecognizer) -> Any?)? = null

    /**
     * UIGestureRecognizer's state change callback
     * @since 1.2.5
     */
    var stateListener: ((UIGestureRecognizer, State?, State?) -> Unit)? = null

    /**
     * @return The current recognizer internal state
     * @since 1.0.0
     */
    var state: State? = null
        protected set(value) {
            logMessage(Log.INFO, "setState: ${this.state?.name} --> ${value?.name}")

            val oldValue = field
            val changed = this.state != value || value == State.Changed
            field = value

            if (changed) {
                stateListener?.invoke(this, oldValue, value)
                val iterator = mStateListeners.listIterator()
                while (iterator.hasNext()) {
                    iterator.next().onStateChanged(this)
                }
            }
        }

    /**
     * Toggle the recognizer enabled state.
     *
     * Set to false to prevent any motion event
     * to be intercepted by this recognizer
     * @since 1.0.0
     */
    var isEnabled: Boolean = true
        set(value) {
            if (field != value) {
                field = value
                if (!value) {
                    reset()
                }
            }
        }

    /**
     * Returns the time (in ms) when the user originally pressed down to start a stream of position events
     * @since 1.2.5
     */
    @Suppress("unused")
    val downTime: Long
        get() = mDownTime

    /**
     * @return Returns the X computed as the location of the original down event.
     * @since 1.2.5
     */
    val downLocationX: Float get() = mDownLocation.x

    /**
     * @return Returns the Y computed as the location of the original down event.
     * @since 1.2.5
     */
    val downLocationY: Float get() = mDownLocation.y

    /**
     * @return Returns the X computed as the location in a given view of the gesture represented by the receiver.
     * @since 1.0.0
     */
    open val currentLocationX: Float get() = mCurrentLocation.x

    /**
     * @return Returns the Y computed as the location in a given view of the gesture represented by the receiver.
     * @since 1.0.0
     */
    open val currentLocationY: Float get() = mCurrentLocation.y

    /**
     * A Boolean value affecting whether touches are delivered to a view when a gesture is recognized
     * @see <a href='https://developer.apple.com/reference/uikit/uigesturerecognizer/1624218-cancelstouchesinview'>cancelstouchesinview</a>
     *
     * @since 1.0.0
     */
    var cancelsTouchesInView: Boolean = true

    /**
     * custom object the instance should keep
     * @since 1.0.0
     */
    var tag: Any? = null

    var id: Long = generateId()
        protected set

    /**
     * Creates a dependency relationship between the receiver and another gesture recognizer when the objects are created
     * @see <a href='https://developer.apple.com/reference/uikit/uigesturerecognizer/1624203-require'>1624203-require</a>
     *
     * @since 1.0.0
     */
    var requireFailureOf: UIGestureRecognizer? = null
        set(other) {
            field?.removeOnStateChangeListenerListener(this)
            field = other
        }

    /**
     * Returns the last recorded event
     * @since 1.0.0
     */
    var lastEvent: MotionEvent? = null
        protected set(mLastEvent) {
            mLastEvent?.recycle()
            field = mLastEvent
        }

    val context: Context? get() = mContextRef.get()

    /**
     * @return Returns the number of touches involved in the gesture represented by the receiver.
     * @since 1.0.0
     */
    open val numberOfTouches: Int get() = mNumberOfTouches

    enum class State {
        Possible,
        Began,
        Changed,
        Failed,
        Cancelled,
        Ended
    }

    private fun generateId(): Long {
        return id++
    }

    @SuppressLint("HandlerLeak")
    protected inner class GestureHandler(mainLooper: Looper) : Handler(mainLooper) {
        override fun handleMessage(msg: Message) {
            this@UIGestureRecognizer.handleMessage(msg)
        }
    }

    open fun reset() {
        state = null
        stopListenForOtherStateChanges()
        setBeginFiringEvents(false)
        removeMessages()
    }

    /**
     * @return Has began firing events
     */
    open fun hasBeganFiringEvents(): Boolean {
        return mBeganFiringEvents
    }

    protected fun setBeginFiringEvents(value: Boolean) {
        mBeganFiringEvents = value
    }

    protected abstract fun removeMessages()

    protected fun removeMessages(vararg messages: Int) {
        for (message in messages) {
            mHandler.removeMessages(message)
        }
    }

    @Suppress("unused")
    protected fun hasMessages(vararg messages: Int): Boolean {
        for (message in messages) {
            if (mHandler.hasMessages(message)) {
                return true
            }
        }
        return false
    }

    internal fun clearStateListeners() {
        mStateListeners.clear()
    }

    protected fun fireActionEvent() {
        actionListener?.invoke(this)
    }

    protected fun addOnStateChangeListenerListener(listener: OnGestureRecognizerStateChangeListener) {
        if (!mStateListeners.contains(listener)) {
            mStateListeners.add(listener)
        }
    }

    protected fun removeOnStateChangeListenerListener(listener: OnGestureRecognizerStateChangeListener): Boolean {
        return mStateListeners.remove(listener)
    }

    protected fun hasOnStateChangeListenerListener(listener: OnGestureRecognizerStateChangeListener): Boolean {
        return mStateListeners.contains(listener)
    }

    @SuppressLint("Recycle")
    open fun onTouchEvent(event: MotionEvent): Boolean {
        lastEvent = MotionEvent.obtain(event)

        // action down
        if (event.actionMasked == MotionEvent.ACTION_DOWN) {
            mPreviousDownLocation.set(mDownLocation)
            mDownLocation.set(event.x, event.y)
            mPreviousDownTime = mDownTime
            mDownTime = event.downTime
        }

        // compute current location
        mNumberOfTouches = computeFocusPoint(event, mCurrentLocation)

        logMessage(Log.VERBOSE, "event.action: ${event.action}, focusPoint: $mCurrentLocation")
        return false
    }

    protected fun computeFocusPoint(event: MotionEvent, out: PointF): Int {
        val actionMasked = event.actionMasked
        val count = event.pointerCount
        val pointerUp = actionMasked == MotionEvent.ACTION_POINTER_UP
        val skipIndex = if (pointerUp) event.actionIndex else -1
        // Determine focal point
        var sumX = 0f
        var sumY = 0f
        for (i in 0 until count) {
            if (skipIndex == i) {
                continue
            }
            sumX += event.getX(i)
            sumY += event.getY(i)
        }

        val div = if (pointerUp) count - 1 else count
        out.x = sumX / div
        out.y = sumY / div
        return if (pointerUp) count - 1 else count
    }

    @Suppress("unused")
    @Throws(Throwable::class)
    protected fun finalize() {
    }

    protected abstract fun handleMessage(msg: Message)

    fun inState(vararg states: State): Boolean {
        return states.contains(state)
    }

    protected fun stopListenForOtherStateChanges() {
        requireFailureOf?.removeOnStateChangeListenerListener(this)
    }

    protected fun listenForOtherStateChanges() {
        requireFailureOf?.addOnStateChangeListenerListener(this)
    }

    override fun toString(): String {
        return javaClass.simpleName + "[state: " + state + ", tag:" + tag + "], touches: $numberOfTouches"
    }

    protected fun logMessage(level: Int, fmt: String) {
        if (!sDebug) {
            return
        }
        Log.println(level, LOG_TAG, "[${javaClass.simpleName}] $fmt")
    }

    @Suppress("unused")
    @TargetApi(Build.VERSION_CODES.CUPCAKE)
    companion object {

        const val VERSION = BuildConfig.VERSION_NAME

        val LOG_TAG: String = UIGestureRecognizer::class.java.simpleName

        /**
         * @return the instance id
         * @since 1.0.0
         */
        var id = 0
            private set

        protected var sDebug = false

        const val TIMEOUT_DELAY_MILLIS = 5
        val LONG_PRESS_TIMEOUT = ViewConfiguration.getLongPressTimeout().toLong()
        val TAP_TIMEOUT = ViewConfiguration.getTapTimeout().toLong()
        val DOUBLE_TAP_TIMEOUT = ViewConfiguration.getDoubleTapTimeout().toLong()
        const val TOUCH_SLOP = 8
        const val DOUBLE_TAP_SLOP = 100
        const val DOUBLE_TAP_TOUCH_SLOP = TOUCH_SLOP

        var logEnabled: Boolean
            get() = sDebug
            set(value) {
                sDebug = value
            }
    }
}
